### PR TITLE
gopages: fix rendering of doclinks for go 1.19 comments

### PR DESCRIPTION
**Pull request by phsym**
Hi ! Thanks for this awesome gopages tool. I love it already.

I followed your recent note about go 1.19 and improved formatting, however I realized that doclinks where not rendered at all so I took the chance to try to improve your tool.

To achieve that I did 2 things :
- I upgraded the godoc library to the latest. ~~By the way I also upgraded other libs but maybe I was wrong, just tell me and I'll revert. I haven't updated the go version to 1.19 in go.mod. Not sure if it's needed~~
- I overloaded the `comment_html` template func the same way you did for the `node_html` template func so that it properly rewrite the doclinks. For some reason, unlike node_html links, doclinks are not prefixed by /pkg

I tried it on my codebase (sorry I can't share the full result) but here's a quick example

Without the change:
image

With the change
image

With the link properly sending to `http://<host>:<port>/pkg/<module>/<package>/#New`

DocLinks to doc from other modules or stdlib redirect to pkg.go.dev

### CODE REVIEW

**Comment by phsym** [comment]
Hum ok, maybe I shouldn't have update all dependencies, some may not be go 1.16 compatible anymore, let me try to fix that

**Comment by phsym** [comment]
Looks like latest golang.org/x/tools requires a version golang.org/x/sys which itself requires `unsafe.Slice` from stdlib which has been added in go 1.17.
Would it be OK for you to drop go 1.16 compatibility ?

**Comment by phsym** [comment]
I removed go 1.16 from the github actions ci and it passed

File: gopages/go.mod [hunk 1]
```diff
@@ -9,8 +9,7 @@ require (
 	github.com/johnstarich/go/pipe v0.2.0
 	github.com/pkg/errors v0.9.1
 	github.com/stretchr/testify v1.5.1
-	golang.org/x/mod v0.6.0-dev.0.20220106191415-9b9b3d81d5e3
-	golang.org/x/net v0.0.0-20220421235706-1d1ef9303861
-	golang.org/x/sys v0.0.0-20220328115105-d36c6a25d886 // indirect
-	golang.org/x/tools v0.1.10
+	golang.org/x/mod v0.7.0
+	golang.org/x/net v0.3.0
+	golang.org/x/tools v0.4.0
```
Comment (JohnStarich) [inline gopages/go.mod:14]: I think dropping 1.16 support for `gopages` is okay, especially since it appears to be required to use the latest godoc support. Can you bump up the supported Go version `go 1.16` accordingly, above?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -23,7 +23,6 @@ jobs:
         platform:
         - ubuntu-latest
         go:
-          - 1.16.x
```
Comment (JohnStarich) [inline .github/workflows/ci.yml:26]: I may need to think about this CI change a bit more. It'd be ideal if we could just skip 1.16 only for `gopages`, that way all the other tools are still tested against 1.16. That darn `x/sys` (indirect) dependency is always causing trouble across Go versions.

File: gopages/internal/generate/generate.go [hunk 3]
```diff
@@ -158,7 +158,7 @@ var crawlerStaticPipe = pipe.New(pipe.Options{}).
 	}).
 	Append(func(args docsArgs) (docsArgs, error) {
 		// Generate main index to redirect to actual content page. Important to separate from 'lib' top-level dir.
-		return args, util.WriteFile(args.FS, filepath.Join(args.OutputPath, "index.html"), []byte(redirect("pkg/"+args.ModulePackage)), scrapeFilePermission)
+		return args, util.WriteFile(args.FS, filepath.Join(args.OutputPath, "index.html"), []byte(redirect("pkg/"+args.ModulePackage+"/")), scrapeFilePermission)
```
Comment (JohnStarich) [inline gopages/internal/generate/generate.go:161]: This can cause extra redirects (`/foo/` -> `/foo`) on some HTTP servers. Is this change required to fix your code's doc links? If so, do you have an example scenario to help reproduce it?

File: gopages/internal/generate/template.go [hunk 4]
```diff
@@ -138,3 +139,21 @@ func nodeHTML(original nodeHTMLFunc, baseURL string, modulePackage string) nodeH
 		return replacer.Replace(original(info, node, linkify))
 	}
 }
+
+type commentHTMLFunc = func(info *godoc.PageInfo, comment string) string
+
+func commentToHTML(original commentHTMLFunc, baseURL string, modulePackage string) commentHTMLFunc {
+	var (
```
Comment (JohnStarich) [inline gopages/internal/generate/template.go:146]: Looks like the body of this function is nearly identical to `nodeHTML`. Let's deduplicate a little. Maybe something like this?

```go
func rewriteLinks(rawHTML, baseURL, modulePackage string) string { ... }

func nodeHTML(original nodeHTMLFunc, baseURL, modulePackage string) nodeHTMLFunc {
	return func(info *godoc.PageInfo, node interface{}, linkify bool) string {
		rawHTML := original(info, node, linkify)
		return rewriteLinks(rawHTML, baseURL, modulePackage)
	}
}

func commentToHTML(original commentHTMLFunc, baseURL string, modulePackage string) commentHTMLFunc {
	return func(info *godoc.PageInfo, comment string) string {
		rawHTML := original(info, comment)
		return rewriteLinks(rawHTML, baseURL, modulePackage)
	}
}
```

**Review by JohnStarich — CHANGES_REQUESTED** [review]
Hey @phsym, thanks for your PR!

Do you have an example (ideally a test if you have time) where this failed before?

**Comment by phsym** [comment]
@JohnStarich here's a quick example where it fails before this PR. I'll look at how to make this a test case
```go
// Package example is a dummy package showing how comments doc is rendered.
//
// Here are some random doc links:
//   - Std library: [os.File] [*os.File] [encoding/xml.Encoder]
//   - Std imported: [json.Decoder] [json.Encoder]
//   - Packages: [os] [json] [encoding/xml]
package example

import (
	"encoding/json"
)

// JSONFunc does random stuff with [json.Decoder] & [json.Encoder]
func JSONFunc() {
	json.Marshal("foobar") // Just to import json
}
```
Not all doc-links are rendered as links, and those which are links to localhost/<package> which eads to a 404
image

After PR, links are properly rendered, a linked to pkg.go.dev when they are external dependencies (internal links are linked properly too, will add them to my test case)
image

File: gopages/go.mod (hunk 1 above)
Comment (phsym) [reply gopages/go.mod:14]: Done

File: gopages/internal/generate/template.go (hunk 4 above)
Comment (phsym) [reply gopages/internal/generate/template.go:149]: Done

File: gopages/internal/generate/generate.go (hunk 3 above)
Comment (phsym) [reply gopages/internal/generate/generate.go:161]: All package & subpackage links are relative links.

When I browse to root `/<base>/`, I get redirected to `/<base>/<mymodule>`. From that page, when I click on a package I'm sent to `/<base>/<mypackage>` instead of `/<base>/<mymodule>/<mypackage>`.
Forcing a trailing slash fixes that issue.

If a server redirects from `/foo/` to `/foo` then it would always render pages with broken links.

I totally agree it would be much better if the server I'm deploying too was configured to perform such redirection, unfortunately I'm not managing this server and cannot change its config. I can revert this change.

Another approach would be to change all the package links to make them absolute instead of relative.

**Comment by phsym** [comment]
I added a test case in gopages/internal/generate/generate_test.go

File: .github/workflows/ci.yml (hunk 2 above)
Comment (phsym) [reply .github/workflows/ci.yml:26]: Do you need something from me on this ?

File: gopages/internal/generate/generate.go (hunk 3 above)
Comment (phsym) [reply gopages/internal/generate/generate.go:161]: Thinking about the extra-redirects removing the trailing slash. The servers doing that are likely to serve pages with broken links anyway. And all links to subpackages have a trailing slash already. So doing sunch kind of redirect in the root index.html would be more consistent with how package  links are rendered evrywhere else

File: .github/workflows/ci.yml (hunk 2 above)
Comment (JohnStarich) [reply .github/workflows/ci.yml:26]: I think this is okay after all. Gopages will have a higher Go compatibility requirement than the others, it's just a bummer we can't run the tests (easily) for all modules on 1.16.

File: gopages/internal/generate/generate_test.go [hunk 5]
```diff
@@ -154,9 +172,97 @@ func Hello() {
 				"src/github.com/my/thing/internal/index.html",
 				"src/index.html",
 			},
+		}, {
```
Comment (JohnStarich) [inline gopages/internal/generate/generate_test.go:175]: tiny nit:
```suggestion
		},
		{
```

**Review by JohnStarich — DISMISSED** [review]
Looks great, thank you @phsym!

**Review by JohnStarich — APPROVED** [review]

File: gopages/internal/generate/generate.go (hunk 3 above)
Comment (JohnStarich) [reply gopages/internal/generate/generate.go:161]: Hmm I'm trying to understand your last comment, though I'm not 100% sure what you mean. Do you have a docs site up that triggers this issue I could take a poke at? Or maybe a minimal example code with which gopages generates the issue?
If I'm understanding correctly, sounds like the package index URLs are badly formed – just not sure I understand how it would happen.

I'd be happy to explore this further, too. Maybe open a new issue to discuss?

**Comment by JohnStarich** [comment]
I've released `v0.1.20` with your change 🎉
